go-fuzz-build/main.go

```diff
@@ -42,503 +42,540 @@ func makeTags() string {
+
+	// Stop if any package had errors.
+	if packages.PrintErrors(initial) > 0 {
+		c.failf("typechecking of %v failed", pkg)
 	}
+
+	c.pkgs = initial
+
+	// Set pkgpath to fully resolved package path.
+	c.pkgpath = initial[0].PkgPath
 }
 
-func createMeta(lits map[Literal]struct{}, blocks []CoverBlock, sonar []CoverBlock) string {
-	meta := MetaData{Blocks: blocks, Sonar: sonar}
-	for k := range lits {
-		meta.Literals = append(meta.Literals, k)
-	}
-	data, err := json.Marshal(meta)
+// loadStd finds the set of standard library package paths.
+func (c *Context) loadStd() {
+	// Find out what packages are in the standard library.
+	stdpkgs, err := packages.Load(nil, "std")
 	if err != nil {
-		failf("failed to serialize meta information: %v", err)
+		c.failf("could not load standard library: %v", err)
 	}
-	f := tempFile()
-	writeFile(f, data)
-	return f
+	c.std = dependencies(stdpkgs)
 }
 
-func buildInstrumentedBinary(pkg string, deps map[string]bool, lits map[Literal]struct{}, blocks *[]CoverBlock, sonar *[]CoverBlock) string {
+// makeWorkdir creates the workdir, logging as requested.
+func (c *Context) makeWorkdir() {
+	// TODO: make workdir stable, so that we can use cmd/go's build cache?
+	// See https://github.com/golang/go/issues/29430.
 	var err error
-	workdir, err = ioutil.TempDir("", "go-fuzz-build")
+	c.workdir, err = ioutil.TempDir("", "go-fuzz-build")
 	if err != nil {
-		failf("failed to create temp dir: %v", err)
+		c.failf("failed to create temp dir: %v", err)
 	}
 	if *flagWork {
-		fmt.Printf("workdir: %v\n", workdir)
-	} else {
-		defer func() {
-			os.RemoveAll(workdir)
-			workdir = ""
-		}()
-	}
-
-	if deps["runtime/cgo"] {
-		// Trick go command into thinking that it has up-to-date sources for cmd/cgo.
-		cgoDir := filepath.Join(workdir, "goroot", "src", "cmd", "cgo")
-		mkdirAll(cgoDir)
-		src := "// +build never\npackage main\n"
-		writeFile(filepath.Join(cgoDir, "fake.go"), []byte(src))
-	}
-	copyDir(filepath.Join(GOROOT, "pkg", "tool"), filepath.Join(workdir, "goroot", "pkg", "tool"), true, nil)
-	if _, err := os.Stat(filepath.Join(GOROOT, "pkg", "include")); err == nil {
-		copyDir(filepath.Join(GOROOT, "pkg", "include"), filepath.Join(workdir, "goroot", "pkg", "include"), true, nil)
+		fmt.Printf("workdir: %v\n", c.workdir)
+	}
+}
+
+// cleanup ensures a clean exit. It should be called on all (controllable) exit paths.
+func (c *Context) cleanup() {
+	if !*flagWork && c.workdir != "" {
+		os.RemoveAll(c.workdir)
+	}
+	if c.cpuprofile != nil {
+		pprof.StopCPUProfile()
+		c.cpuprofile.Close()
+	}
+}
+
+// populateWorkdir prepares workdir for builds.
+func (c *Context) populateWorkdir() {
+	// TODO: instead of reconstructing the world,
+	// can we use a bunch of replace directives in a go.mod?
+
+	// TODO: make all this I/O concurrent (up to a limit).
+	// It's a non-trivial part of build time.
+	// Question: Do it here or in copyDir?
+
+	// TODO: consider using hard links for
```

Comment:
I think this can't work across mounts, and /tmp is frequently a different mount.

Reply:
Ack. Will remove that comment in a follow-up.